Add `django-framework` extension and `django-framework` init profile

weiiwang01 opened the pull request:
Add the `django-framework` that can package a Django project into a ROCK, similar to the `flask-framework`.
Some refactoring was performed, and some improvements were made over the `flask-framework` extension as well.

Changes include:

- Update the extension part name prefix from `flask/` to `flask-framework/`.
- Install `statsd-exporter` along with the Gunicorn server.
- Install `ca-certificates` in the ROCK image by default.
- Disable the WSGI path check when `services.flask.command` or `services.django.command` is overridden.
- Improve the quality of generated parts.

-----
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

weiiwang01 commented on rockcraft/extensions/gunicorn.py:
```diff
@@ -0,0 +1,340 @@
+    def gen_install_app_part(self) -> Dict[str, Any]:
+        """Return the prime list for the Flask project."""
+        if "django-framework/install-app" not in self.yaml_data.get("parts", {}):
+            return {
+                "plugin": "dump",
+                "source": ".",
+                "organize": {self.name: "django/app"},
+                "stage": ["django/app"],
+                "prime": ["django/app"],
+            }
```
I am not very confident about this. Is there a better way to achieve the same effect? I tried this, but it doesn't work.

```yaml
django-framework/install-app:
  plugin: dump
  source: "{self.name}"
  organize: 
    ".": django/app
```

weiiwang01 replied:
Does the wildcard patten here matches files start with dot `.`?

tigarmo replied:
hmm I'm not sure. If it doesn't, then maybe this will:

```
  organize:
    "*": django/app
    ".*": django/app

weiiwang01 replied:
I did some experiment and found that `organize` indeed uses `glob` rules (which don't match files starting with `.`) instead of `fnmatch` rules. Here is the final working version (the target directory needs to end with `/`):

```yaml
  organize:
    "*": django/app/
    ".*": django/app/
```

Thank you for your suggestion! updated.

weiiwang01 commented on docs/how-to/use-django-extension.rst:
Updated, thanks!

weiiwang01 commented on docs/how-to/use-django-extension.rst:
```diff
@@ -0,0 +1,69 @@
+Using the django-framework extension
+------------------------------------
+
+The Flask extension is compatible with the ``bare`` and ``ubuntu@22.04`` bases.
```
Added, thanks!

weiiwang01 commented on docs/how-to/use-django-extension.rst:
```diff
@@ -0,0 +1,69 @@
+
+Example:
+
+.. code-block:: yaml
```
Gotcha, I have replaced all example blocks with real examples and added some tests for them.

weiiwang01 commented on docs/how-to/use-flask-extension.rst:
Yes, I have searched and replaced all instances of "ROCK"

weiiwang01 commented on rockcraft/extensions/gunicorn.py:
```diff
@@ -0,0 +1,340 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2023 Canonical Ltd.
```
Updated, thanks!

weiiwang01 commented on rockcraft/extensions/gunicorn.py:
```diff
@@ -0,0 +1,340 @@
+    @override
+    def get_supported_bases() -> Tuple[str, ...]:
+        """Return supported bases."""
+        return "bare", "ubuntu@20.04", "ubuntu@22.04", "ubuntu:20.04", "ubuntu:22.04"
```
No, I forgot to update here, updated, thanks!

tigarmo reviewed (commented):
Should we have a spread test that actually stands up a Django app and interacts with it in some manner?

weiiwang01 commented on rockcraft/commands/init.py:
```diff
@@ -118,6 +118,47 @@ class InitCommand(AppCommand):
+                name: {name}
+                base: ubuntu@22.04 # the base environment for this Django application
+                version: '0.1' # just for humans. Semantic versioning is recommended
+                summary: A summary of your Flask application # 79 char long summary
```
Updated, thanks!

tigarmo commented:
> Should we have a spread test that actually stands up a Django app and interacts with it in some manner?

bump @weiiwang01

weiiwang01 commented:
> > Should we have a spread test that actually stands up a Django app and interacts with it in some manner?
> 
> bump @weiiwang01

Sorry, I missed this comment. I've added a new spread test for `django-extension` similar to the one for `flask-extension`.

tigarmo commented:
@weiiwang01 as we discussed I moved the `feature/12f` branch to the current `main`. Feel free to close this PR and open a new one, or update this one, it's up to you. Thanks

weiiwang01 commented:
> @weiiwang01 as we discussed I moved the `feature/12f` branch to the current `main`. Feel free to close this PR and open a new one, or update this one, it's up to you. Thanks

Thanks! since there isn't too much conflict, I've updated this PR and cleaned up the commits.

tigarmo commented on tests/spread/__init__.py:
why are these `__init__.py` files here? These aren't Python packages

weiiwang01 replied:
It must be automatically created by the IDE, I have removed those files, thanks!

tigarmo commented on tests/spread/general/extension-django/task.yaml:
```diff
@@ -0,0 +1,34 @@
+summary: django extension test
```
this is very nice 👍🏾

lengau commented on docs/how-to/code/use-django-extension/example/example_django/example_django/asgi.py:
```diff
@@ -0,0 +1,16 @@
+
+from django.core.asgi import get_asgi_application
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "example_django.settings")
```
Is this the preferred way to do this? It seems to me like putting this inside the rock's environment would be cleaner (though IDK what downsides it would have).

weiiwang01 replied:
Yes, all Django projects in the tests are generated using the Django admin tool, so I think it's okay.

lengau commented on docs/how-to/code/use-django-extension/task.yaml:
```diff
@@ -0,0 +1,21 @@
+###########################################
+summary: test the "Use the django-framework extension" guide
+
+environment:
+  ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS: "1"
+
+execute: |
+  cd example
+  rockcraft pack --verbosity debug
+  cd ../override_example
+  rockcraft pack --verbosity debug
+  cd ..
+
+restore: |
+  rm -rf */*.rock
```
```suggestion
environment:
  ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS: "1"
  EXAMPLE_DIR/example: example
  EXAMPLE_DIR/override_example: override_example

execute: |
  cd $EXAMPLE_DIR
  rockcraft pack --verbosity debug

restore: |
  cd $EXAMPLE_DIR
  rm *.rock
  rockcraft clean
```

This uses [variants](https://github.com/snapcore/spread?tab=readme-ov-file#variants) to break the test up. That way if `example` starts failing, it'll independently check whether `override_example` still succeeds.

weiiwang01 replied:
Learned, thanks!

lengau commented on tests/spread/general/extension-django/task.yaml:
```diff
@@ -0,0 +1,34 @@
+summary: django extension test
+
+execute: |
+  export ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS=true
```
```suggestion

environment:
  ROCKCRAFT_ENABLE_EXPERIMENTAL_EXTENSIONS: "true"

execute: |
```

lengau reviewed (approved):
Looks good overall. I made some minor suggestions, but they're definitely not blockers for this PR.

tigarmo commented:
Spread failures are unrelated (I'm working on them)
